Support SAF properly

Stypox opened the pull request:
<!-- Hey there. Thank you so much for improving NewPipe, and filling out the details. Having roughly the same layout helps everyone considerably :)-->

#### What is it?
- [x] Bugfix (user facing)
- [ ] Feature (user facing)
- [x] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
This is the rebased #3777, and also contains:
- Setting migration that enables SAF everywhere possible.

Other than that, it is identical to #3777 (by @wb9688):
1. Use SAF or NoNonsense-FilePicker *everywhere* based on the setting.
2. Don't have `requestLegacyExternalStorage` anymore in the manifest.
3. Allow only NoNonsense-FilePicker for versions below Android 5 and allow only SAF starting at Android 10.
4. Move `SharpInputStream` into NewPipe and add `SharpOutputStream`.
5. Show folder picker in download dialog if no download folder set yet.

#### TODO
- [x] "SAF" in the UI should be replaced with something more easily understandable for non-Android devs, like "system folder picker"
- [x] Remove "some devices are not compatible" from SAF setting description
- [x] Investigate whether renaming the zip works - it works
- [x] ~~Keeping the setting in the Download menu doesn't make sense anymore~~ implementing a miscellaneous menu is left for another PR
- [x] Investigate "NewPipe saf Leaks" - I can't understand why the name changed, but it's not important
- [x] Explain SharpOutputStream
- [x] "filename.mp4 (1)" is created instead of "filename (1).mp4" - fixed by passing mime type to file picker
- [x] Fix strange behaviour when app loses access to saf download folder
- [x] Reimplement storing import/export path

#### Fixes the following issue(s)
Fixes #3465, fixes #5198, and perhaps some others.

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

wb9688 committed:
Support SAF properly

wb9688 committed:
Move Stored(File|Directory)Helper into NewPipe

wb9688 committed:
Use file picker for export DB

wb9688 committed:
Fix non-SAF actions

wb9688 committed:
Only ask for storage permissions when not using SAF

Stypox committed:
Add setting migration to promote using SAF

opusforlife2 commented:
Wow you're on fire.

opusforlife2 commented:
https://github.com/TeamNewPipe/NewPipe/pull/3777#issuecomment-702796747?

Stypox committed:
Fix ContentSettingsManager tests

Stypox commented:
@opusforlife2 I don't know, with the setting migration everyone should seamlessly switch to SAF, but I'd keep the option anyway since on some phones Files apps can be difficult to use, and some users might complain. The code wouldn't be simplified much if we were to remove the setting.

opusforlife2 commented:
Viva la democracy.

Stypox commented:
###### The tests fail because of some type issues and of https://github.com/TeamNewPipe/NewPipe/pull/5225#issuecomment-760478351, I will fix them later. So I'm providing a **manually-built APK**.
Could you confirm that all of the following things work? in this APK: [app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5817057/app-debug.zip)

- Export database
- Import database
- Export subscriptions
- Import subscriptions from previous export
- Import subscriptions from YouTube
- Setting default audio download path
- Setting default video download path
- Downloading audio
- Downloading video
- Downloading audio when "Ask where to download" is enabled
- Downloading video when "Ask where to download" is enabled

Thanks in advance.

@imcmjha @xibr @B0pol @Domiuns @fajis1 @mqus @MD77MD @Poolitzer @mhmdanas

mhmdanas commented:
@Stypox I tested all the things you listed and faced no issues.

fajis1 commented:
everything you asked me to look at worked, but there were some crashes that I had. I reported them via email.

When I was in subscriptions it would crash when I tried to leave the subscription area. 

when I tried to play the videos and audio I downloaded from within the app download area. It worked for the files that auto downloaded to the specified folder. But if it asked me for the folder those would cause the app to crash when I tried to play the videos or audio. 

The videos and audio played just fine when I found them with the file explorer etc. So they did download successfully.

xibr commented:
## Exception
* __User Action:__ ui error
* __Request:__ App crash, UI failure
* __Content Country:__ KR
* __Content Language:__ ko
* __App Language:__ en
* __Service:__ none
* __Version:__ 0.20.8
* __OS:__ Linux samsung/star2ltexx/star2lte:10/QP1A.190711.020/G965FXXUCFTK1:user/release-keys 10 - 29
<details><summary><b>Crash log </b></summary><p>

```
java.lang.IllegalStateException: Fragment no longer exists for key f0: unique id ccfa4eee-12c7-47ba-8199-8c5cab8ca937
	at androidx.fragment.app.FragmentManager.getFragment(FragmentManager.java:772)
	at androidx.fragment.app.FragmentStatePagerAdapterMenuWorkaround.restoreState(FragmentStatePagerAdapterMenuWorkaround.java:311)
	at androidx.viewpager.widget.ViewPager.onRestoreInstanceState(ViewPager.java:1461)
	at android.view.View.dispatchRestoreInstanceState(View.java:21574)
	at android.view.ViewGroup.dispatchRestoreInstanceState(ViewGroup.java:4693)
	at android.view.ViewGroup.dispatchRestoreInstanceState(ViewGroup.java:4699)
	at android.view.View.restoreHierarchyState(View.java:21552)
	at androidx.fragment.app.Fragment.restoreViewState(Fragment.java:573)
	at androidx.fragment.app.FragmentStateManager.restoreViewState(FragmentStateManager.java:356)
	at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1189)
	at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1356)
	at androidx.fragment.app.FragmentManager.moveFragmentToExpectedState(FragmentManager.java:1434)
	at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1497)
	at androidx.fragment.app.BackStackRecord.executePopOps(BackStackRecord.java:510)
	at androidx.fragment.app.FragmentManager.executeOps(FragmentManager.java:2166)
	at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:1992)
	at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:1947)
	at androidx.fragment.app.FragmentManager.popBackStackImmediate(FragmentManager.java:658)
	at androidx.fragment.app.FragmentManager.popBackStackImmediate(FragmentManager.java:569)
	at androidx.fragment.app.FragmentManager.handleOnBackPressed(FragmentManager.java:541)
	at androidx.fragment.app.FragmentManager$1.handleOnBackPressed(FragmentManager.java:350)
	at androidx.activity.OnBackPressedDispatcher.onBackPressed(OnBackPressedDispatcher.java:192)
	at androidx.activity.ComponentActivity.onBackPressed(ComponentActivity.java:314)
	at org.schabi.newpipe.MainActivity.onBackPressed(MainActivity.java:591)
	at android.app.Activity.onKeyUp(Activity.java:3686)
	at android.view.KeyEvent.dispatch(KeyEvent.java:3578)
	at android.app.Activity.dispatchKeyEvent(Activity.java:4003)
	at androidx.core.app.ComponentActivity.superDispatchKeyEvent(ComponentActivity.java:122)
	at androidx.core.view.KeyEventDispatcher.dispatchKeyEvent(KeyEventDispatcher.java:84)
	at androidx.core.app.ComponentActivity.dispatchKeyEvent(ComponentActivity.java:140)
	at androidx.appcompat.app.AppCompatActivity.dispatchKeyEvent(AppCompatActivity.java:569)
	at androidx.appcompat.view.WindowCallbackWrapper.dispatchKeyEvent(WindowCallbackWrapper.java:59)
	at androidx.appcompat.app.AppCompatDelegateImpl$AppCompatWindowCallback.dispatchKeyEvent(AppCompatDelegateImpl.java:3054)
	at androidx.appcompat.view.WindowCallbackWrapper.dispatchKeyEvent(WindowCallbackWrapper.java:59)
	at com.android.internal.policy.DecorView.dispatchKeyEvent(DecorView.java:624)
	at android.view.ViewRootImpl$ViewPostImeInputStage.processKeyEvent(ViewRootImpl.java:6492)
	at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:6356)
	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5796)
	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:5849)
	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:5815)
	at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:5971)
	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:5823)
	at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:6028)
	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5796)
	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:5849)
	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:5815)
	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:5823)
	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5796)
	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:5849)
	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:5815)
	at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:6004)
	at android.view.ViewRootImpl$ImeInputStage.onFinishedInputEvent(ViewRootImpl.java:6173)
	at android.view.inputmethod.InputMethodManager$PendingEvent.run(InputMethodManager.java:3826)
	at android.view.inputmethod.InputMethodManager.invokeFinishedInputEventCallback(InputMethodManager.java:3308)
	at android.view.inputmethod.InputMethodManager.finishedInputEvent(InputMethodManager.java:3299)
	at android.view.inputmethod.InputMethodManager$ImeInputEventSender.onInputEventFinished(InputMethodManager.java:3803)
	at android.view.InputEventSender.dispatchInputEventFinished(InputEventSender.java:143)
	at android.os.MessageQueue.nativePollOnce(Native Method)
	at android.os.MessageQueue.next(MessageQueue.java:336)
	at android.os.Looper.loop(Looper.java:197)
	at android.app.ActivityThread.main(ActivityThread.java:8167)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)

```
</details>
<hr>

xibr commented:
Everything works perfectly, I don't know if this crash is related.

Go to subscriptions and then back the app crashes.
screen recorder.

https://user-images.githubusercontent.com/20098154/104659290-ca7ec980-56dd-11eb-8fb0-121ce28251bc.mp4

Stypox commented:
I fixed the tests, thank you @XiangRongLin for suggesting me to Sync gradle ;-)
Thank you for testing! Yeah, the random crashes when going back are caused by #5414 and fixed in #5417, so no need to worry about those. This is the apk built by the Ci: https://github.com/TeamNewPipe/NewPipe/suites/1834544416/artifacts/35447435

mqus commented:
Everything works for me (Android 4.4.2 (Kitkat, API19), just two things:
- I couldn't test import from youtube(no subscriptions there), but I don't think there will be any errors there.
- The SAF option in the settings is disabled and greyed out, though I assume that is intentional.

mhmdanas commented:
@mqus yes, that is intentional.

Stypox commented:
@dkraft @blauertee @k9janer @bingoxo @illustribe @RickyM7 I am also pinging you for testing since you have Android 11, by looking at the issues. Could you test the apk [above](https://github.com/TeamNewPipe/NewPipe/pull/5415#issuecomment-760486053) with the [steps above](https://github.com/TeamNewPipe/NewPipe/pull/5415#issuecomment-760486053)?

RickyM7 commented:
> @dkraft @blauertee @k9janer @bingoxo @illustribe @RickyM7 I am also pinging you for testing since you have Android 11, by looking at the issues. Could you test the apk above with the steps above?

In fact the version of android I use is 7.0, you must have been confused with the version of MIUI I use that is 11. Sorry!

Stypox commented:
@RickyM7 oh, sorry ;-)
If you have some time to test anyway, I would appreciate it, so that we can make sure this works everywhere

illustribe commented:
Sorry, no A11 here :/

RickyM7 commented:
@Stypox Of course!

RickyM7 commented:
@Stypox Ok, I tested what I could and everything is working perfectly. The only thing I couldn't test was "Import subscriptions from YouTube" because I don't have an account.

- Device: Redmi Note 4 (MIUI 11 - Android 7.0)

Fs00 commented:
If I may add, I think that "SAF" in the UI should be replaced with something more easily understandable for non-Android devs, like "system folder picker".
Also, is it correct to say in settings that "some devices are not compatible" with SAF, given that this PR enables it by default? :slightly_smiling:

Stypox commented:
> If I may add, I think that "SAF" in the UI should be replaced with something more easily understandable for non-Android devs, like "system folder picker".

You are right, good idea

> Also, is it correct to say in settings that "some devices are not compatible" with SAF, given that this PR enables it by default?

No, that has to be removed now. Before this PR there were incompatibilities, but now there shouldn't be anymore.

opusforlife2 commented:
Tested everything except "Import subscriptions from YouTube". Werks. 🎉 

Exporting database with current release (so NNFP) doesn't let you rename the zip, but this PR does. Intentional? Wasn't the ability to rename the export zip removed in an earlier version for some reason?

Before this PR, SAF was limited to downloads. But now it works with various other IO operations, so keeping the setting in the Download menu doesn't make sense anymore. But no other menu suits it either. Time for a new Miscellaneous menu? :D

Aside: Other debug apks show "Leaks" as an option, but your apk shows "NewPipe saf Leaks". How did that happen?

XiangRongLin commented on app/src/main/java/org/schabi/newpipe/streams/io/SharpInputStream.java:
```diff
@@ -0,0 +1,48 @@
+import java.io.IOException;
+import java.io.InputStream;
+
+public class SharpInputStream extends InputStream {
```
Could you add an explanation what the difference of this and `SharpOutputStream` to normal java streams is and why it is needed. The `SharpStream` class only tells me that it's from C#

Stypox replied:
Ok, I will try to investigate it myself, since I don't know ;-)

wb9688 replied:
SharpStream is a wrapper around the 2 different APIs for SAF and non-SAF. It has both input and output like in C#, while in Java those are usually different classes. Because a class could only extend one class, there is SharpInputStream and SharpOutputStream that wraps SharpStream while extending InputStream or OutputStream, so that you could use it with built-in Java stuff that supports InputStream or OutputStream.

XiangRongLin commented on app/src/main/java/org/schabi/newpipe/settings/ContentSettingsManager.kt:
```diff
@@ -17,8 +19,9 @@ class ContentSettingsManager(private val fileLocator: NewPipeFileLocator) {
-        ZipOutputStream(BufferedOutputStream(FileOutputStream(outputPath)))
+    fun exportDatabase(preferences: SharedPreferences, file: StoredFileHelper) {
+        file.create()
+        ZipOutputStream(BufferedOutputStream(SharpOutputStream(file.stream)))
```
@wb9688 @Stypox Does that then mean, that the SharpOutputStream can be removed here?

wb9688 replied:
No, SharpOutputStream wraps file.stream (which is a SharpStream) so that it is possible to use it there.

XiangRongLin replied:
Ah now i understood it. thanks for clarifying

XiangRongLin commented on app/src/main/java/org/schabi/newpipe/util/ZipHelper.java:
```diff
@@ -59,24 +62,23 @@ public static void addFileToZip(final ZipOutputStream outZip, final String file,
                                              final String name) throws Exception {
         try (ZipInputStream inZip = new ZipInputStream(new BufferedInputStream(
-                new FileInputStream(filePath)))) {
+                new SharpInputStream(zipFile.getStream())))) {
```
same here

Poolitzer commented:
since the others checked everything else, I confirmed that importing from youtube works

k9janer commented:
So far everything works on my Pixel 3a with GrapheneOS based on Android 11.

Just 4 things I have noticed:

- Downloading video/audio while a video is playing and with the option "Ask where to download" enabled opens the save dialogue but then it immediately asks for permission to "Display over other apps" as NewPipe switches to a different player. You have to press back a few times to go back to the save dialogue.

- When you download a file twice with the "Ask where to download" option enabled, the new file will be named "filename.mp4 (1)". (The .mp4 part isn't at the end of the file). NewPipe crashes when you try to open this download and other apps on the device won't recognize this file as a video file.

- This might be an Android/GrapheneOS bug: When you set up a default save location for video/audio my Pixel 3a automatically points to the folder "Download/NewPipe" I created previously. However there is no "use this folder" option. I have to go back to "Download" and re-enter the "NewPipe" folder for the "use this folder" option to appear.

- NewPipe behaves a bit weird when you clear storage access:

I tried it a few different times and for the first few tries, NewPipe just gave the error message "couldn't create folder" or something like that and I could fix it by going to the settings and re-registering the folder.

But right now it's behaving differently: I'm in a situation where NewPipe recognizes that storage access was cleared and it displays the toast "no download folder set yet, choose the default downloader folder now". It then opens to "use this folder" dialogue and downloads the file perfectly fine. The problem is: According to the Android App Info, NewPipe now has access to the folder I choose in this dialogue, but the same error toast and procedure keeps repeating every time I try to download something.

Stypox committed:
Improve SAF switch descriptions in settings

Stypox committed:
Add comments to SharpStreams

Stypox committed:
Pass mime type so that SAF treats file extension correctly

Stypox committed:
Fix strange behaviour when app loses access to saf download folder

Stypox commented:
> but then it immediately asks for permission to "Display over other apps" as NewPipe switches to a different player

That's expected and will happen only once, so I wouldn't consider this something that needs fixing

> When you download a file twice with the "Ask where to download" option enabled, the new file will be named "filename.mp4 (1)". (The .mp4 part isn't at the end of the file). NewPipe crashes when you try to open this download and other apps on the device won't recognize this file as a video file.

Fixed, thank you for the report! Now if you want to overwrite the file you have to manually click on the old file in the system picker, otherwise "filename (1).mp4" will be created. The latter behaviour is not under our control, that's just how the system picker works, anyway both cases are now handled correctly by NewPipe.

> This might be an Android/GrapheneOS bug: When you set up a default save location for video/audio my Pixel 3a automatically points to the folder "Download/NewPipe" I created previously. However there is no "use this folder" option. I have to go back to "Download" and re-enter the "NewPipe" folder for the "use this folder" option to appear.

Yeah, this is not something under our control unfortunately

> NewPipe behaves a bit weird when you clear storage access

Fixed, thank you again for the explanation :-D

k9janer commented:
> > but then it immediately asks for permission to "Display over other apps" as NewPipe switches to a different player
> 
> That's expected and will happen only once, so I wouldn't consider this something that needs fixing
> 

Okay, I understand that this issue really isn't the focus of this PR but I want to explain why I think this dialogue is pretty annoying:

a) I dislike the popup player and have no use for it on my device. For this reason I don't want to give NewPipe this permission which means I get this permission dialogue every time I want to download something as long as "ask where to download" is enabled.
b) In this moment I/the user wants to just quickly download something. The permission dialogue is completely unexpected and there is no explanation why it suddenly appears. Since this permission could be abused by apps and the OS warns users against granting this permission, some users could think that NewPipe has some shady intentions.

However I understand that it's outside of the scope of this PR. So let's move on.

> > This might be an Android/GrapheneOS bug: When you set up a default save location for video/audio my Pixel 3a automatically points to the folder "Download/NewPipe" I created previously. However there is no "use this folder" option. I have to go back to "Download" and re-enter the "NewPipe" folder for the "use this folder" option to appear.
> 
> Yeah, this is not something under our control unfortunately
> 

Okay, that's what I thought. I don't have this issue on my FireHD10 tablet so I assumed it was an OS bug/weirdness.

> > NewPipe behaves a bit weird when you clear storage access
> 
> Fixed, thank you again for the explanation :-D Mind testing again? @k9janer
> 

Okay, I retested the two issues again and I can say:

- File name: The file name is now correctly created as filename (1).mp4. 
- Clear storage access: NewPipe now works mostly correct in this regard. The only - extremely minor/artificial - issue I faced is that NewPipe doesn't properly check if a file still exists when "ask where to download is disabled":
1. Download a video file
2. Go to the file manager of your choice and delete the video file you just downloaded
3. Switch back to NewPipe and download the same video file again
Result: NewPipe will now claim "A downloaded file with this name already exists" even after it was deleted.
You can still press on "overwrite" and NewPipe will download the file perfectly fine and even without(!) a (1) at the end of the file name.

Stypox commented:
@k9janer thank you for your further testing!

> Okay, I understand that this issue really isn't the focus of this PR but I want to explain why I think this dialogue is confusing in this situation:

The app just tries to minimize the player to popup whenever it loses focus, since that's the default setting. You can change this in `Settings -> Video & Audio -> Minimize on app switch`

> NewPipe will now claim "A downloaded file with this name already exists" even after it was deleted.

Oh, this is a nasty issue caused by how the downloader currently works: it is unable to detect if a downloaded file was deleted. I don't think it can be fixed easily, and anyway was not introduced by this PR, as (I am pretty sure) this also exists on the current release.

Poolitzer commented:
> this also exists on the current release.

it does not

Stypox committed:
Check if file really exists before asking user to overwrite it

Stypox commented:
@TobiGr @B0pol I rebased once again. I think this has been tested again and again by multiple users and is ready to be merged. Could you review it?

> NewPipe will now claim "A downloaded file with this name already exists" even after it was deleted.
>
> > this also exists on the current release.
>
> it does not

Fixed in b9a41033eb9f5a1e7df24482db57585e80812bea (even though I still don't understand why it works in the current release). @Poolitzer could you confirm?

TobiGr commented:
The
 ```
    /*//////////////////////////////////////////////////////////////////////////
    // LifeCycle
    //////////////////////////////////////////////////////////////////////////*/
```
comment in line 162 in DownloadDialog is misplaced. I cannot comment on that line, because you did not modify it.

Stypox committed:
Remove misplaced comment

Stypox commented:
@TobiGr fixed

Poolitzer commented:
> @Poolitzer could you confirm?

since then, I switched to a pixel 4a, the SAF requirement is there as well right? Or should I get the old phone out for this?

Stypox commented:
Use the 4a, no problem

Stypox commented:
@TobiGr I rebased again. Could you review? As I said above it has been tested again and again, so it should be safe to merge after code has been reviewed

Poolitzer commented:
Downloaded the artifact from the CI, ran the following tests, on a Pixel 4a 5g:

- [x] Export database
- [x]     Import database
- [x]     Export subscriptions
- [x]     Import subscriptions from previous export
- [x]     Import subscriptions from YouTube
- [x]     Setting default audio download path
- [x]     Setting default video download path
- [x]     Downloading audio
- [x]     Downloading video
- [x]     Downloading audio when "Ask where to download" is enabled
- [x]     Downloading video when "Ask where to download" is enabled

I found one bug. I downloaded a file, deleted it with the file app, then redownloaded the same file, and I get the "error" that _a downloaded file with this name already exists_. I can override it, so its fine, but still a bug.

Stypox commented:
> I found one bug. I downloaded a file, deleted it with the file app, then redownloaded the same file, and I get the "error" that a downloaded file with this name already exists. I can override it, so its fine, but still a bug.

@Poolitzer I can't reproduce, could you better explain which steps you have followed? This should have already been fixed by 32a717c6c66870d8f6b0f0fc69e55263d89cd1a8

Poolitzer commented:
How can I make sure that the apk I downloaded included the commit? Before I try to explain an already fixed bug to you ;P

Stypox commented:
[This](https://github.com/TeamNewPipe/NewPipe/suites/2408549478/artifacts/51298199) apk works for me

k9janer commented:
The check for an existing file works perfectly now n my Pixel 3a.

NewPipe properly detects whether a file already exists or not. 👍

Poolitzer commented:
> This apk works for me

Ha, still doesn't for me.

download is set to ask.

I searched for _short video_ and selected [this](https://www.youtube.com/watch?v=H9154xIoYTA).
I hit DOwnload from the button menu below the video, this is set on Audio (doesn't matter though). Then I select ok, and the DCIM folder.
After the download finishes, I swap over to the Files application (the one from google). I do not close the NewPipe app. Then I delete the file from the folder.
I swap back to NewPipe, redownload the file, its properly gone from the select folder scren, but after pressing ok it gives the error.

Hope thats clear enough :O

opusforlife2 commented:
This is the exact bug I reported (#3782), which was fixed (#3794), but for the old no-nonsense file picker. :D Strangely, I noted there that this bug didn't occur on SAF back then. xD

Stypox commented:
@Poolitzer thank you, now I could reproduce. I was missing the "Ask" option. Now it should be fixed, could you test again?

Poolitzer commented:
I tested, it still gives me the same error, but a weird new behavior. When I delete the file from the file manager, it is gone, as well in NewPipe, but after hitting ok in the select folder... (window? are they called windows? or screen?) thingy, a 0 B file with the same name as the one I want to download gets generated, and then I get the error. So the error is actually correct, because now a file exists, it just shouldn't ;P

Again, I leave NewPipe open when deleting the file, I think that is important.

Stypox commented:
What the ...? "a 0 B file with the same name as the one I want to download gets generated" is what happened before... I'm gonna test again, maybe I pushed the wrong commit

Poolitzer commented:
Maybe I downloaded the wrong apk, I take the one from the build and test run.

Stypox committed:
Revert all commits related to ContentSettingsFragment

Revert "Annotate methode parameters as NonNull"
This reverts commit 004907d306f84bbe800b624e9af650f73989652b.

Revert "Commit path immediately when import backup"
This reverts commit 05eb0d0fbe7ffc13d6b9a431d0825c6ce6c8dc79.

Revert "Set ImportExportDataPath only on successful import"
This reverts commit f13a1b04e64d01d2337662b724ddff4b9ede54fc.

Revert "Set ImportExportDataPath only on successful export"
This reverts commit fd4408e572f5c5aacf5c74ace91ef7c1ed46c474.

Revert "Invert if condition in ContentSettingsFragment.setImportExportDataPath for better readability"
This reverts commit 92ab9cae27620337d21e4f634b918a919aa5d830.

Revert "Move ContentSettingsFragment.isValidPath to helpers and add unit test for it."
This reverts commit fa2b11b7685d989d1e28e62f65ff1f80d890a9d1.

Revert "Save backup import/export location for feature import/exports"
This reverts commit 82f43ac6a64c1754e023aed6c5cd7a117f1a99dc.

Remove FilePathHelperTest file

Stypox committed:
Reimplement storing backup import/export path

#6319 and #6402 were reverted before adding SAF changes, and have been readded at the end of SAF changes

Stypox commented:
I rebased again:
- The first commit reverts #6319 and #6402 (storing the import/export directory) and the last one reimplements them from scratch. Hopefully I did not mess up anything. @XiangRongLin @ATofighi please review the last commit ("Reimplement storing backup import/export path").
- I tested on API 19 emulated (FilePicker) and on my Android 10 phone (SAF) and everything seems to work fine.
- @Poolitzer I tried following again the exact steps you point out at https://github.com/TeamNewPipe/NewPipe/pull/5415#issuecomment-812692830, and I can't reproduce. Could you test one last time with the latest APK in the CI? Sorry for keeping bothering you :-) Anyway, even if the issue is still there for you, I'd merge this anyway, as it is a minor problem that shouldn't come up with the average user and in any case does not cause crashes or problematic behaviour.
- @TobiGr please please please let's merge this as soon as possible, or it will become even more difficult to rebase ;-)

XiangRongLin commented:
> The first commit reverts #6319 and #6402 (storing the import/export directory) and the last one reimplements them from scratch. Hopefully I did not mess up anything. XiangRongLin ATofighi please review the last commit ("Reimplement storing backup import/export path").
I tested on API 19 emulated (FilePicker) and on my Android 10 phone (SAF) and everything seems to work fine.

Seems correct to me

Poolitzer commented:
@Stypox I updated my phone in the meantime to Android 11, but it works now. Can't rule out if its a software bug sadly, but the app works as intended \o/

TobiGr reviewed (dismissed):
@Poolitzer Perfect. 

Let's merge this once 0.21.4 is released

TobiGr commented:
Oh no!
We've got a merge conflict

Stypox commented:
Rebased @TobiGr

Stypox commented:
Finally :-D :-D
Thank you @Poolitzer and everyone else for the tests you did and thank you @wb9688 for the initial work :-)
